source/common/websocket/codec.cc

```diff
@@ -0,0 +1,219 @@
+#include "source/common/websocket/codec.h"
+
+#include <algorithm>
+#include <array>
+#include <cstdint>
+#include <memory>
+#include <vector>
+
+#include "source/common/buffer/buffer_impl.h"
+#include "source/common/common/scalar_to_byte_vector.h"
+
+namespace Envoy {
+namespace WebSocket {
+
+absl::optional<std::vector<uint8_t>> Encoder::encodeFrameHeader(const Frame& frame) {
+  if (std::find(kFrameOpcodes.begin(), kFrameOpcodes.end(), frame.opcode_) == kFrameOpcodes.end()) {
+    ENVOY_LOG(debug, "Failed to encode websocket frame with invalid opcode: {}", frame.opcode_);
+    return absl::nullopt;
+  }
+  std::vector<uint8_t> output;
+  // Set flags and opcode
+  pushScalarToByteVector(
+      static_cast<uint8_t>(frame.final_fragment_ ? (0x80 | frame.opcode_) : frame.opcode_), output);
+
+  // Set payload length
+  if (frame.payload_length_ <= 125) {
+    // Set mask bit and 7-bit length
+    pushScalarToByteVector(frame.masking_key_ ? static_cast<uint8_t>(frame.payload_length_ | 0x80)
+                                              : static_cast<uint8_t>(frame.payload_length_),
+                           output);
+  } else if (frame.payload_length_ <= 65535) {
+    // Set mask bit and 16-bit length indicator
+    pushScalarToByteVector(static_cast<uint8_t>(frame.masking_key_ ? 0xfe : 0x7e), output);
+    // Set 16-bit length
+    pushScalarToByteVector(htobe16(frame.payload_length_), output);
+  } else {
+    // Set mask bit and 64-bit length indicator
+    pushScalarToByteVector(static_cast<uint8_t>(frame.masking_key_ ? 0xff : 0x7f), output);
+    // Set 64-bit length
+    pushScalarToByteVector(htobe64(frame.payload_length_), output);
+  }
+  // Set masking key
+  if (frame.masking_key_) {
+    pushScalarToByteVector(htobe32(frame.masking_key_.value()), output);
+  }
+  return output;
+}
+
+void Decoder::frameMaskFlag(uint8_t mask_and_length) {
+  num_remaining_masking_key_bytes_ = mask_and_length & 0x80 ? kMaskingKeyLength : 0;
+  length_ = mask_and_length & 0x7F;
+}
+
+void Decoder::frameDataStart() {
+  frame_.payload_length_ = length_;
+  if (length_ == 0) {
+    state_ = State::FrameFinished;
+  } else {
+    if (decode_payload_) {
+      frame_.payload_ = std::make_unique<Buffer::OwnedImpl>();
+    }
+    state_ = State::FramePayload;
+  }
+}
+
+void Decoder::frameData(const uint8_t* mem, uint64_t length) { frame_.payload_->add(mem, length); }
+
+void Decoder::frameDataEnd(absl::optional<std::vector<Frame>>& output) {
+  output->push_back(std::move(frame_));
+  resetDecoder();
+}
```

Comment:
If the output is optional, should we check it before call the `push_back`? Or the output cannot be optional?

Reply:
It can be optional. It was initially as the way like you request now. We have changed it in the review. Now it is initialised to an empty vector.
```
absl::optional<std::vector<Frame>> output = std::vector<Frame>();
```
and do the empty check in the return to make it optional.
```
return !output->empty() ? std::move(output) : absl::nullopt;
```
this will avoid multiple empty check calls.

Reply:
If the ouput parameter here always has value, may be just use `std::vector<Frame>&` here as the parameter type.